js/private/js_devserver.js

```diff
@@ -0,0 +1,161 @@
+const path = require('path')
+const fs = require('fs')
+const os = require('os')
+const child_process = require('child_process')
+
+const RUNFILES_ROOT = path.join(
+    process.env.RUNFILES,
+    process.env.JS_BINARY__WORKSPACE
+)
+
+const synced = new Map()
+
+async function mkdirp(p) {
+    if (p && !fs.existsSync(p)) {
+        await mkdirp(path.dirname(p))
+        await fs.promises.mkdir(p)
+    }
+}
+
+function isExecutable(stats) {
+    return (
+        !!(stats.mode & fs.constants.S_IXUSR) ||
+        !!(stats.mode & fs.constants.S_IXGRP) ||
+        !!(stats.mode & fs.constants.S_IXOTH)
+    )
+}
+
+async function syncRecursive(src, dst) {
+    try {
+        const lstat = await fs.promises.lstat(src)
+        const last = synced.get(src)
+        if (!lstat.isDirectory() && last && lstat.mtimeMs == last) {
+            // this file is already up-to-date
+            return
+        }
+        synced.set(src, lstat.mtimeMs)
+        const exists = fs.existsSync(dst)
+        if (lstat.isSymbolicLink()) {
+            console.log(
+                `Syncing symlink ${src.slice(RUNFILES_ROOT.length + 1)}`
+            )
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.symlink(src, dst)
+        } else if (lstat.isDirectory()) {
+            await mkdirp(dst)
+            const contents = await fs.promises.readdir(src)
+            await Promise.all(
```

Comment:
seems like this code is more synchronous than required?

Reply:
Yeah. I need to be initially synchronous for the initial sync but then it can be more asynchronous for incremental syncs. I'll think about how to make that work.

Reply:
Too synchronous because it does one level of the tree at a time before recursing a level deeper? 🤔

Reply:
Actually if I'm recalling the `data_files` correctly I don't think we even need recursion and can have one giant `Promise.all`?
```
function sync(...) { return Promise.all(files.map(... syncRecursive(src, dst)) }

function syncRecursive(src, dst) {
   ...if/lstat/check-cache...

   await mkdirp(path.dirname(dst));
   if (symlink) ... link ...
   else if (file) ... copy/chmod ...
   else if (directory) ... mkdir() ...    // <--- no "p"
}
```

Note that the results of `syncRecursive` are put in `Promise.all` and not individual `await`s so it should be 100% parallel?